### PR TITLE
Support additional function returning PossiblyNormalCompletion

### PR DESCRIPTION
**Pull request by cblappert**
Release Notes: Allow additional functions to return PossiblyNormalCompletions.

Similar to the logic in `Program.js`, we need to emit an `if` statement at the end of an additional function that may return or throw.

In `functions.js` we process the PossiblyNormalCompletion, joining the consequent and alternate effects and creating a set of args + buildnode for emitting the conditional throw. Then where we would originally serialize/visit the return value, we will instead apply the joined effects and emit the conditional throw/return.

### CODE REVIEW

File: src/serializer/functions.js [hunk 1]
```diff
@@ -126,11 +127,27 @@ export class Functions {
...
-    return {
-      effects,
-      transforms: [],
-    };
+    let [result, generator] = effects;
+    let retValue: AdditionalFunctionEffects = { effects, transforms: [] };
+    // Create the effects, arguments and buildNode for the return value, saving them in AdditionalFunctionEffects
+    if (result instanceof PossiblyNormalCompletion) {
+      let { joinCondition, consequent, alternate, consequentEffects, alternateEffects } = result;
+      let joinedEffects = Join.joinEffects(this.realm, joinCondition, consequentEffects, alternateEffects, true);
```
Comment (hermanventer) [inline src/serializer/functions.js]: I don't understand why you are doing the join at this point. It does not seem like a natural join point to me. In particular, exceptions does not get swallowed at this point, so they should not be joining their effects with the normal effects here. Let's discuss this face to face.

File: src/serializer/ResidualHeapSerializer.js [hunk 2]
```diff
@@ -1803,7 +1807,10 @@ export class ResidualHeapSerializer {
       invariant(object instanceof ObjectValue);
       this._emitProperty(object, propertyBinding.key, propertyBinding.descriptor, true);
     }
-    invariant(result instanceof Value, "TODO: support PossiblyNormalCompletion return from additional function");
+    invariant(
+      result instanceof Value || result instanceof PossiblyNormalCompletion,
+      "TODO: support non-value non-PossiblyNormalCompletion return from additional function"
```
[the remark this answers is not in the document]
Comment (cblappert) [reply src/serializer/ResidualHeapSerializer.js]: I can change the comment to say that AbruptCompletions mean the function unconditionally terminates in an exception in which case we will issue a compiler diagnostic and FatalError earlier on.

**Comment by cblappert** [comment]
It seems like fixing this to apply abrupt effects only for serializing the throw and apply the normal effects only for serializing the return would require significant changes to additional functions. This is something that should be done to emit more succinct code sometime in the future.

**Comment by hermanventer** [comment]
Still waiting for a response to my last question.

**Pull request closed without merging**